src/Autofac/Features/OpenGenerics/OpenGenericServiceBinder.cs

```diff
@@ -86,24 +92,33 @@ internal static class OpenGenericServiceBinder
         [NotNullWhen(returnValue: true)] out Func<IComponentContext, IEnumerable<Parameter>, object>? constructedFactory,
         [NotNullWhen(returnValue: true)] out Service[]? constructedServices)
     {
-        if (serviceWithType.ServiceType.IsGenericType && !serviceWithType.ServiceType.IsGenericTypeDefinition)
+        var serviceWithTypeServiceType = serviceWithType.ServiceType;
+        if (serviceWithTypeServiceType.IsGenericType && !serviceWithTypeServiceType.IsGenericTypeDefinition)
         {
-            var definitionService = (IServiceWithType)serviceWithType.ChangeType(serviceWithType.ServiceType.GetGenericTypeDefinition());
-            var serviceGenericArguments = serviceWithType.ServiceType.GetGenericArguments();
+            var definitionService = (IServiceWithType)serviceWithType.ChangeType(GetGenericTypeDefinition(serviceWithTypeServiceType));
```

Comment:
Sorry, I did a fast pass at this before - I noticed that there is some type-related caching going on now with this `GenericTypeDefinitionByType` dictionary.

I think that's going to break [the new work that got done to isolate caching so we can unload it and allow AssemblyLoadContexts to be unloaded](https://github.com/autofac/Autofac/blob/2e622bda29521333fa4b6d93466dfbc768f77a0e/src/Autofac/Core/Lifetime/LifetimeScope.cs#L208)

If there's a cache of type info it'll need to be [stored in the `ReflectionCacheSet`](https://github.com/autofac/Autofac/blob/develop/src/Autofac/Core/ReflectionCacheSet.cs) so it can be unloaded if needed.

For example, [we cache all the `IEnumerable<T>` we figure out](https://github.com/autofac/Autofac/blob/86c36ab0a70b0f1f43e7821b5aea73cf9a4e8136/src/Autofac/Util/InternalTypeExtensions.cs#L148) so we can unload those references. This seems similar. I'm unclear if we have tests that cover open generics unloading in isolated scopes since there wasn't any caching like this before.

It may warrant [a handy internal accessor](https://github.com/autofac/Autofac/blob/86c36ab0a7/src/Autofac/Core/InternalReflectionCaches.cs) to make it fast to retrieve.

Reply:
done